sdk/iptables/iptables.go

```diff
@@ -107,14 +111,27 @@ func Setup(cfg Config) error {
 		cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", ProxyOutputRedirectChain, "-p", "tcp", "-j", "REDIRECT", "--to-port", strconv.Itoa(cfg.ProxyOutboundPort))
 
 		// The DNS rules are applied before the rules that directs all TCP traffic, so that the traffic going to port 53 goes through this rule first.
-		if cfg.ConsulDNSIP != "" {
+		if cfg.ConsulDNSIP != "" && cfg.ConsulDNSPort == 0 {
 			// Traffic in the DNSChain is directed to the Consul DNS Service IP.
 			cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", DNSChain, "-p", "udp", "--dport", "53", "-j", "DNAT", "--to-destination", cfg.ConsulDNSIP)
 			cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", DNSChain, "-p", "tcp", "--dport", "53", "-j", "DNAT", "--to-destination", cfg.ConsulDNSIP)
 
 			// For outbound TCP and UDP traffic going to port 53 (DNS), jump to the DNSChain.
 			cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", "OUTPUT", "-p", "udp", "--dport", "53", "-j", DNSChain)
 			cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", "OUTPUT", "-p", "tcp", "--dport", "53", "-j", DNSChain)
+		} else if cfg.ConsulDNSPort != 0 {
+			consulDNSIP := "127.0.0.1"
+			if cfg.ConsulDNSIP != "" {
+				consulDNSIP = cfg.ConsulDNSIP
+			}
+			consulDNSHostPort := fmt.Sprintf("%s:%d", consulDNSIP, cfg.ConsulDNSPort)
+			// Traffic in the DNSChain is directed to the Consul DNS Service IP.
+			cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", DNSChain, "-p", "udp", "-d", consulDNSIP, "--dport", "53", "-j", "DNAT", "--to-destination", consulDNSHostPort)
+			cfg.IptablesProvider.AddRule("iptables", "-t", "nat", "-A", DNSChain, "-p", "tcp", "-d", consulDNSIP, "--dport", "53", "-j", "DNAT", "--to-destination", consulDNSHostPort)
```

Comment:
Disclaimer: I know very little about `iptables` so this might be a silly question! 😅 

It looks to me like setting both `ConsulDNSIP` and `ConsulDNSPort` causes us to pass the given IP as the `-d` (destination) flag here, which we don't do if only `ConsulDNSIP` is given.

Is that correct? Wouldn't that only redirect packets already destined for the given `ConsulDNSIP`, when presumably we want to redirect all packets on port 53?

Reply:
Yep that's correct. Prior to this release, we allowed to pass consul dns IP, but now we also want to pass a port. So I kept previous behavior for backward compatibility.